features/237-cms-validation/api.md

```diff
@@ -0,0 +1,146 @@
+- Field validation is executed before form validation.
+
+## Configuration
+
+- One validation item is a dictionary which contains the validation configuration (depends on where it is placed).
+  - Each validation optionally contains a `message` field, which would be the error message if that validation item cannot pass.
+  - Otherwise, a default error message would be shown.
+- `validation` may be a validation item or contains a list of validation items.
+
+```
+new:
+  # page validation
+  validation:
+    - expression: xxx
+    - expression: xxx
+  fields:
+    - name: xxx
+      type: xxx
+      # field validation
+      validation:
+        pattern: /^xxx$/
+        message:
+```
+
+## Expression
+
+- `expression` is the base type of validation, supported by all validation.
+- An expression is a string, which can be evaluated to give boolean value.
+- All other types of validation are predifined expression, i.e. they are all compiled to expression at runtime.
+- Expression evaluation would be based on this library https://github.com/joewalnes/filtrex, while CMS would provide extra predefined functions for common use case.
+
+### Functions
+
+- Passing wrong data type to the function would throw error in validation.
+
+### Expression context: `value`
+
+Developer uses the variable `value` in the expression to get the value from the field or form.
+
+- Field validation
+  - `value`: value of the field
+  - e.g. `value in ('abc', 'def')`
+- Form validation
+  - `value`: value of the form
+  - `get(value: formValue, fieldName: string)`
+  - e.g. `get(value, "full_name") != get(value, "nick_name")`
+
+#### Predefined functions
+
+- String
+  - `length(value: string)`
+  - `upper(value: string)`
+  - `lower(value: string)`
+  - `substring(value: string, from: number, to: number)`
+- Number
+- Datetime
+  - `datetime(str: string)`
+  - `timestamp(data: datetime | string)`
+    - e.g. `timestamp(value) > timestamp('2018-01-01')`
+- JSON
+  - `get(data: json, key: string | number)`
+    - return the value at key
+    - e.g. `get(get(value, "a"), "b")`, `{"a": {"b": 1}}` => `1`
+    - e.g. `get(get(value, "a"), 0)`, `{"a": [1]}` => `1`
+  - `hasKey(data: json, key: string | number)`
+    - return if the key exist
+- Location
+- References
+  - `length(value: reference[])`
+- EmbeddedReference
+  - `get(data: embeddedReference, key: string)`
+- EmbeddedReferences
+  - `length(value: embeddedReference[])`
+  - `get(data: embeddedReference[], index: number)`
+- Asset
+  - `size(value: asset)`
+
+## Predefined validation for field validation
+
+### `required`
+
+types: `String`, `Datetime`, `References`, `EmbeddedReferences`, `Asset`
+
+```
+validation:
+- required: true
+# String / References / EmbeddedReferences equivalent to
+- expression: length() > 0
+# Datetime / Asset equivalent to
+- expression: value != null
+```
+
+Note that, `required: false` is no-op.
+
+### `pattern`
```

Comment:
is it regular expression? If yes call it `regex` (etc) instead of `pattern`?

Reply:
What about both `regex` and `pattern`, while `pattern` can be something like `email`, `creditcard` and `url`, which are some predefined regex.

Reply:
good idea.